src/sciline/sphinxext/domain_types.py

```diff
@@ -0,0 +1,153 @@
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright (c) 2024 Scipp contributors (https://github.com/scipp)
+"""A Sphinx extension for improved typehints rendering.
+
+`sphinx-autodoc-typehints <https://github.com/tox-dev/sphinx-autodoc-typehints>`_
+implements formatting for typehints but often results in hard-to-read annotations
+for :class:`typing.NewType`.
+As those are ubiquitous when using Sciline, this extension improves formatting.
```

Comment:
```suggestion
As those may be ubiquitous when using Sciline, this extension improves formatting.
```
`NewType` might just be how we currently use it.